src/node_url.cc

```diff
@@ -407,29 +407,64 @@ int64_t ParseIPv4Number(const char* start, const char* end) {
   return strtoll(start, nullptr, R);
 }
 
+// https://url.spec.whatwg.org/#ipv4-number-parser
+bool IsIPv4NumberValid(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  // If a number starts with '0' it might be a number with base 8 or base
+  // 16. If not, checking if all characters are digits proves that it is a
+  // base 10 number.
+  if (input.size() >= 2 && input[0] == '0') {
+    if (input[1] == 'X' || input[1] == 'x') {
+      if (input.size() == 2) {
+        return true;
+      }
+
+      return input.find_first_not_of("0123456789abcdefABCDEF", 2) ==
+             std::string_view::npos;
+    }
+
+    return input.find_first_not_of("01234567", 1) == std::string_view::npos;
+  }
+
+  return std::all_of(input.begin(), input.end(), ::isdigit);
+}
+
 // https://url.spec.whatwg.org/#ends-in-a-number-checker
-bool EndsInANumber(const std::string& input) {
-  std::vector<std::string> parts = SplitString(input, '.', false);
+inline bool EndsInANumber(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  char delimiter = '.';
+  std::string_view::iterator pointer_start = input.begin();
+  std::string_view::iterator pointer_end = input.end();
 
-  if (parts.empty()) return false;
+  uint8_t parts_size = std::count(pointer_start, pointer_end, delimiter);
```

Comment:
Why uint8_t?

Reply:
Because it's an IPv4, so there shouldn't be more parts than, lets say, 4. In short.. it would always be a small number (I think uint8_t still quite big for this)
but yeah.. maybe I should change to what std::count returns so it don't mess up the number..
Do you think it should be changed?

Reply:
uint8_t is correct. referencing spec:

```
A valid IPv4-address string must be four shortest possible strings of [ASCII digits]
(https://infra.spec.whatwg.org/#ascii-digit), representing a decimal number in the 
range 0 to 255, inclusive, separated from each other by U+002E (.).
```

Reply:
> maybe I should change to what std::count returns so it don't mess up the number..

I think that’s a good idea, yeah. Reading this code it’s really not obvious that anything _in the code_ guarantees that the return value fits into an `uint8_t`, even if that’s a type that can hold the all _valid_ results. As an extreme case, if `input` contains 256 dots, `parts_size` would be 0 here, and I assume we don’t want to treat that the same as a string that doesn’t contain any dots.